File: epi/IsStringPalindromicPunctuation.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
public class IsStringPalindromicPunctuation {
  @EpiTest(testDataFile = "is_string_palindromic_punctuation.tsv")

  public static boolean isPalindrome(String s) {
    /*s = s.toLowerCase();
    int i = 0, j = s.length() - 1;

    while(i < j) {
      char head = s.charAt(i);
      if(!isAlphaNumeric(head)) {
        i++;
        continue;
      }

      char tail = s.charAt(j);
      if(!isAlphaNumeric(tail)) {
        j--;
        continue;
      }

      if(head != tail) {
        return false;
      }
      i++;
      j--;
    }

    return true;*/
    //Empty return true?, upper and Lower case mixed, if we only need to take care the alpha numeric
    //Is Case sensitive
    //T O(N) S O(1)
    /*int len = s.length(), left = 0, right = len - 1;

    while(left < right) {
      char lCh = s.charAt(left), rCh = s.charAt(right);
      if(!isAlphaNumeric(lCh)) {
        left++;
        continue;
      }

      if(!isAlphaNumeric(rCh)) {
        right--;
        continue;
      }

      if(lowerCase(lCh) != lowerCase(rCh)) {
        return false;
      }
      left++;
      right--;
    }

    return true;*/
    return epi.kt.IsStringPalindromicPunctuation.INSTANCE.isPalindrome(s);
  }

  private static char lowerCase(char ch) {
    if(ch >= 'A' && ch <= 'Z') {
      return (char) (ch - 'A' + 'a');
    } else {
      return ch;
    }
  }

  private static boolean isAlphaNumeric(char ch) {
    if((ch >= '0' && ch <= '9')
    || (ch >= 'a' && ch <= 'z')
    || (ch >= 'A' && ch <= 'Z')) {
      return true;
    } else {
      return false;
    }
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "IsStringPalindromicPunctuation.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi/kt/IsStringPalindromicPunctuation.kt
package epi.kt

object IsStringPalindromicPunctuation {
    fun isPalindrome(s: String): Boolean {
        var lS = toLowerCase(s)
        val len = lS.length
        var left = 0
        var right = len - 1

        while(left < right) {
            if(!isAlphaNumeric(lS[left])) {
                left++
                continue
            }

            if(!isAlphaNumeric(lS[right])) {
                right--
                continue
            }

            if(lS[left] != lS[right]) {
                return false
            }

            left++
            right--
        }

        return true
    }

    private fun toLowerCase(s: String): String {
        val sCharArray = s.toCharArray()

        for(i in s.indices) {
            val c = s[i]
            if(c in 'A'..'Z') {
                sCharArray[i] = 'a'.plus(c - 'A')
            }
        }

        return String(sCharArray)
    }

    private fun isAlphaNumeric(ch: Char): Boolean {
        return (ch in '0'..'9') || (ch in 'a'..'z')
    }
}